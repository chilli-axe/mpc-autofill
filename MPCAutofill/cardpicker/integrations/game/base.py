from abc import ABC, abstractmethod
from typing import Any, Callable, Optional, Type
from urllib.parse import urljoin, urlparse

import requests
import sentry_sdk

from cardpicker.models import DFCPair


def default_is_response_valid(response: requests.Response) -> bool:
    return response.status_code == 200


class ImportSite(ABC):
    """
    Abstract base class for an import site integration. These should facilitate importing a list of cards
    based on a supplied URL.
    """

    @staticmethod
    @abstractmethod
    def get_host_names() -> list[str]:
        """
        Returns the host names for this import site, e.g. ["google.com", "www.google.com"].
        """

        ...

    @classmethod
    @abstractmethod
    def retrieve_card_list(cls, url: str) -> str:
        """
        Takes a URL pointing to a card list hosted on this class's site, queries the site's API / whatever for
        the card list, formats it and returns it.
        """

        ...

    @classmethod
    def request(
        cls,
        path: str,
        method: str = "GET",
        is_response_valid: Callable[[requests.Response], bool] = default_is_response_valid,
        netloc: Optional[str] = None,
        headers: Optional[dict[str, Any]] = None,
    ) -> requests.Response:
        url = urljoin(f"https://{netloc or cls.get_host_names()[0]}", path)
        response = requests.request(url=url, method=method, headers=headers)
        if not is_response_valid(response):
            sentry_sdk.capture_message(response.text)
            raise cls.InvalidURLException(url)
        return response

    class InvalidURLException(Exception):
        def __init__(self, url: str):
            super().__init__(
                f"There was a problem with importing your list from {self.__class__.__name__} at URL {url}. "
                f"Check that your URL is correct and try again."
            )


class GameIntegration(ABC):
    """
    Abstract base class for a game integration. These collect code to integrate with a specific card game
    to enrich the app and tailor it more closely to the community's expectations.
    """

    # region abstract methods

    @classmethod
    @abstractmethod
    def get_dfc_pairs(cls) -> list[DFCPair]:
        ...

    @classmethod
    @abstractmethod
    def get_import_sites(cls) -> list[Type[ImportSite]]:
        ...

    # endregion

    @classmethod
    def query_import_site(cls, url: Optional[str]) -> Optional[str]:
        if url is None:
            raise ValueError("No decklist URL provided.")
        netloc = urlparse(url).netloc
        for site in cls.get_import_sites():
            if netloc in site.get_host_names():
                text = site.retrieve_card_list(url)
                cleaned_text = "\n".join(
                    [stripped_line for line in text.split("\n") if len(stripped_line := line.strip()) > 0]
                )
                if len(cleaned_text) > 0:
                    return cleaned_text
        return None


__all__ = ["ImportSite", "GameIntegration"]
