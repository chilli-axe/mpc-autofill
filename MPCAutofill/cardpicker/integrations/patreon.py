import platform
from typing import Optional

import requests

from cardpicker.schema_types import CampaignClass, Supporter, SupporterTier

from MPCAutofill.settings import PATREON_ACCESS, PATREON_URL

# Header must be included to access Patreon info
patreon_header = {
    "Authorization": f"Bearer {PATREON_ACCESS}",
    "User-Agent": f"Patreon-Python, version 0.5.1, platform {platform.platform()}",
}


def get_patreon_campaign_details() -> tuple[Optional[CampaignClass], Optional[dict[str, SupporterTier]]]:
    """
    Get needed patreon campaign details.
    :return: Campaign ID, list of dictionaries containing supporter tier info.
    """

    if not PATREON_URL:
        return None, None

    try:
        res = requests.get(
            # https://docs.patreon.com/#get-api-oauth2-v2-campaigns
            url="https://www.patreon.com/api/oauth2/v2/campaigns",
            params={
                "include": "tiers",
                "fields[campaign]": "summary",
                "fields[tier]": ",".join(["title", "description", "amount_cents"]),
            },
            headers=patreon_header,
        ).json()

        # Properly format campaign details
        campaign = CampaignClass(id=res["data"][0]["id"], about=res["data"][0]["attributes"]["summary"])

        # Properly format campaign tiers
        tiers: dict[str, SupporterTier] = {}
        for tier in res["included"]:
            # Ignore free tier
            if tier["attributes"]["amount_cents"] < 1:
                continue
            # Build dictionary of tiers to reference by ID
            tiers[tier["id"]] = SupporterTier(
                title=tier["attributes"]["title"],
                description=tier["attributes"]["description"],
                usd=round(tier["attributes"]["amount_cents"] / 100),
            )
    except KeyError:
        print("Warning: Cannot locate Patreon campaign. Check Patreon access token!")
        return None, None
    return campaign, tiers


def get_patrons(
    campaign_id: str, campaign_tiers: dict[str, SupporterTier], page: Optional[str] = None
) -> Optional[list[Supporter]]:
    """
    Get our patreon contributors.
    :note: https://docs.patreon.com/#get-api-oauth2-v2-campaigns-campaign_id-members
    :return: List of dictionaries containing patreon contributor info.
    """

    if not PATREON_URL:
        return None

    try:
        # Use page if provided, otherwise build a complete query
        res = (
            requests.get(url=page, headers=patreon_header).json()
            if page
            else requests.get(
                url=f"https://www.patreon.com/api/oauth2/v2/campaigns/{campaign_id}/members",
                params={
                    "include": "currently_entitled_tiers",
                    "fields[member]": ",".join(
                        ["full_name", "campaign_lifetime_support_cents", "pledge_relationship_start", "patron_status"]
                    ),
                },
                headers=patreon_header,
            ).json()
        )

        # Return formatted list of patrons
        results: list[Supporter] = []
        for mem in res.get("data", []):

            # Skip non-active members
            mem_details = mem.get("attributes", {})
            if mem_details.get("patron_status") != "active_patron":
                continue

            # Pull subscribed tiers for this member
            mem_tiers = [
                campaign_tiers[t["id"]]
                for t in mem.get("relationships", {}).get("currently_entitled_tiers", {}).get("data", [])
                if t.get("id") in campaign_tiers
            ]

            # Skip members with no subscribed tiers
            if not mem_tiers:
                continue

            # Use member's highest subscribed tier
            current_tier = sorted(mem_tiers, key=lambda item: item.usd)[0]

            # Add member to results
            results.append(
                Supporter(
                    name=mem_details.get("full_name", "Unknown"),
                    tier=current_tier.title or "Unknown Tier",
                    date=mem_details.get("pledge_relationship_start", "2024-01-01")[:10],
                    usd=current_tier.usd or 5,
                )
            )

        # Check for additional page results
        next_page = res.get("links", {}).get("next")
        if next_page:
            results.extend(get_patrons(campaign_id=campaign_id, campaign_tiers=campaign_tiers, page=next_page) or [])

        # Return sorted results at top-level
        if page:
            return results
        return sorted(results, key=lambda item: item.usd, reverse=True)

    # Unable to retrieve patrons
    except KeyError:
        print("Warning: Cannot locate Patreon campaign. Check Patreon access token!")
        return None


__all__ = ["get_patreon_campaign_details", "get_patrons"]
