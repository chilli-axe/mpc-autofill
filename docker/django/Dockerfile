# Docker to serve Django via Unicorn
FROM ubuntu:22.04 AS builder

# Keeps Python from generating .pyc files in the container
ENV PYTHONDONTWRITEBYTECODE=1
# Turns off buffering for easier container logging
ENV PYTHONUNBUFFERED=1

# Install tools and additional dependencies
RUN apt-get update && \
    apt-get install -y --no-install-recommends \
        dos2unix \
        python3-pip python3-dev \
        gcc netcat curl libpq-dev \
    && rm -rf /var/lib/apt/lists/*

# Copy requirements.txt
COPY MPCAutofill/requirements.txt /MPCAutofill/
WORKDIR /MPCAutofill

# Install pip requirements
RUN pip3 install gunicorn wheel
RUN pip3 install -r requirements.txt

# Copy relevant files from repository
COPY docker /MPCAutofill/docker
COPY MPCAutofill /MPCAutofill/MPCAutofill

# Handle environment variables - write out `.env` with the variables passed from the
# compose file (and overwrite `.env` as copied from the host machine in the process)
RUN touch /MPCAutofill/MPCAutofill/MPCAutofill/.env
RUN echo "DATABASE_HOST=$DATABASE_HOST" > /MPCAutofill/MPCAutofill/MPCAutofill/.env  # using > to overwrite the file
RUN echo "ELASTICSEARCH_HOST=$ELASTICSEARCH_HOST" >> /MPCAutofill/MPCAutofill/MPCAutofill/.env
RUN echo "DEBUG=$DEBUG" >> /MPCAutofill/MPCAutofill/MPCAutofill/.env
RUN echo "ALLOWED_HOSTS=$ALLOWED_HOSTS" >> /MPCAutofill/MPCAutofill/MPCAutofill/.env
RUN echo "GAME=$GAME" >> /MPCAutofill/MPCAutofill/MPCAutofill/.env
RUN echo "MOXFIELD_SECRET=$MOXFIELD_SECRET" >> /MPCAutofill/MPCAutofill/MPCAutofill/.env

# Make sure that all scripts are executable, and in case we
# checked out under Windows with CRLF, convert line endings
RUN chmod +x docker/django/*.sh && \
    find . -type f -name "*.sh" -exec dos2unix {} \;

# Let's be nice to the user and give some clear error message
# if they missed to install any of the mandatory files.
RUN ./docker/django/check_client_secrets.sh && \
    ./docker/django/check_drives.sh

# Create a non-root user and add permission to access the /MPCAutofill folder
RUN adduser -u 5678 --disabled-password --gecos "" mpcautofill && \
    chown -R mpcautofill /MPCAutofill
USER mpcautofill

# Place environment file if one does not already exist
RUN ./docker/django/copy_env_file.sh

# Prepare folder for static files (create now as user for permissions)
RUN mkdir -p static


FROM builder AS webserver

EXPOSE 8000
WORKDIR /MPCAutofill/MPCAutofill
ENTRYPOINT ["/MPCAutofill/docker/django/entrypoint.sh"]
CMD ["gunicorn", "MPCAutofill.wsgi:application", "--bind", "0.0.0.0:8000"]

FROM builder AS worker

WORKDIR /MPCAutofill/MPCAutofill
ENTRYPOINT ["/MPCAutofill/docker/django/entrypoint.sh"]
CMD ["python3", "manage.py", "qcluster"]
