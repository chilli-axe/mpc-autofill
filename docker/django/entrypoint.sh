#!/bin/bash
set -e

# Wait for postgres to come up
echo "Waiting for Postgres..."
sleep 10
while ! nc -z postgres 5432; do
    echo "Still waiting for Postgres..."
    sleep 5
done

# Wait for elasticsearch to come up
echo "Waiting for Elasticsearch..."
sleep 10
until curl --silent --output /dev/null http://elasticsearch:9200/_cat/health?h=st; do
    echo "Still waiting for Elasticsearch..."
    sleep 5
done

# Gather static files
python3 manage.py collectstatic --noinput

# Check if we are running for the first time
if ! python3 manage.py migrate --check; then
    # Run migrations and populate database
    echo "Migrate Django database..."
    python3 manage.py migrate
    echo "Read drives from CSV..."
    python3 manage.py import_sources
    echo "Scan drives and update database..."
    python3 manage.py update_database
    echo "Retrieve double-faced cards from Scryfall..."
    python3 manage.py update_dfcs
fi

exec "$@"
