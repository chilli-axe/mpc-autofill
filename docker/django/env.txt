DJANGO_SECRET_KEY=SehrGeheim123
DJANGO_DEBUG=False
# Uncomment the following line to allow access from other machines.
# DJANGO_ALLOWED_HOSTS="*"
DATABASE_ENGINE=django.db.backends.postgresql
DATABASE_NAME=mpcautofill
ELASTICSEARCH_HOST=elasticsearch
