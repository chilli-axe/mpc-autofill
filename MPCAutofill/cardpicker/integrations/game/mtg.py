import re
from typing import Any, Type
from urllib.parse import parse_qs, urlparse

import ratelimit

from django.conf import settings

from cardpicker.integrations.game.base import GameIntegration, ImportSite
from cardpicker.models import DFCPair
from cardpicker.utils import get_json_endpoint_rate_limited

# region import sites


class Aetherhub(ImportSite):
    @staticmethod
    def get_host_names() -> list[str]:
        return ["aetherhub.com", "www.aetherhub.com"]

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        path = urlparse(url).path
        deck_id = path.split("-")[-1]
        if not deck_id:
            raise cls.InvalidURLException(url)
        return cls.request(path=f"Deck/MtgoDeckExport/{deck_id}").text


class Archidekt(ImportSite):
    @staticmethod
    def get_host_names() -> list[str]:
        return ["archidekt.com", "www.archidekt.com"]

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        path = urlparse(url).path
        regex_results = re.compile(r"^/decks/(.+?)(?:[\#\/].*)?$").search(path)
        if regex_results is None:
            raise cls.InvalidURLException(url)
        deck_id = regex_results.groups()[0]
        if not deck_id:
            raise cls.InvalidURLException(url)
        response_json = cls.request(path=f"api/decks/{deck_id}/").json()
        return "\n".join([f"{x['quantity']} {x['card']['oracleCard']['name']}" for x in response_json["cards"]])


class CubeCobra(ImportSite):
    @staticmethod
    def get_host_names() -> list[str]:
        return ["cubecobra.com", "www.cubecobra.com"]

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        path = urlparse(url).path
        cube_id = path.split("/")[-1]
        response = cls.request(
            path=(
                f"cube/download/plaintext/{cube_id}"
                f"?primary=Color%20Category"
                f"&secondary=Types-Multicolor&tertiary=Mana%20Value"
                f"&quaternary=Alphabetical"
                f"&showother=false"
            ),
            is_response_valid=lambda r: urlparse(r.url).path != "/404",
        )
        # filter out lines like `# mainboard` and `# maybeboard` which were recently introduced by cubecobra
        return "\n".join([x for x in response.text.split("\n") if not x.startswith("# ")])


class Deckstats(ImportSite):
    @staticmethod
    def get_host_names() -> list[str]:
        return ["deckstats.net", "www.deckstats.net"]

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        path = urlparse(url).path
        regex_results = re.compile(r"^/decks/(\d*)/(\d*)").search(path)
        if regex_results is None:
            raise cls.InvalidURLException(url)
        owner_id, deck_id = regex_results.groups()
        if not owner_id or not deck_id:
            raise cls.InvalidURLException(url)
        response = cls.request(
            path=(
                f"api.php"
                f"?action=get_deck"
                f"&id_type=saved"
                f"&owner_id={owner_id}"
                f"&id={deck_id}"
                f"&response_type=list"
            )
        )
        card_list = response.json()["list"]
        for x in [
            "//Main\n",
            "//Sideboard\n",
            "SB: ",
            "//Instant\n",
            "//Sorcery\n",
            "//Land\n",
            " # !Commander",
        ]:
            card_list = card_list.replace(x, "")
        return card_list


class MagicVille(ImportSite):
    @staticmethod
    def get_host_names() -> list[str]:
        return ["magic-ville.com", "www.magic-ville.com"]

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        qs = parse_qs(url)
        if len(qs) != 1 or len(qs[(key := list(qs.keys()).pop())]) != 1:
            raise cls.InvalidURLException(url)
        deck_id = qs[key].pop()
        response = cls.request(path=f"fr/decks/dl_appr?ref={deck_id}&save=1")
        card_list = response.text
        for x in ["// www.magic-ville.com deck file\r\n", "SB: "]:
            card_list = card_list.replace(x, "")
        if not card_list:
            raise cls.InvalidURLException(url)
        return card_list


class ManaStack(ImportSite):
    @staticmethod
    def get_host_names() -> list[str]:
        return ["manastack.com"]  # www. is explicitly not valid

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        path = urlparse(url).path
        deck_id = path.split("/")[-1]
        if not deck_id:
            raise cls.InvalidURLException(url)
        response = cls.request(path=f"api/deck/list?slug={deck_id}")
        response_json = response.json()
        card_list = "\n".join([f"{item['count']} {item['card']['name']}" for item in response_json["list"]["cards"]])
        return card_list


class Moxfield(ImportSite):
    @staticmethod
    def get_host_names() -> list[str]:
        return ["www.moxfield.com", "moxfield.com"]  # moxfield prefers www.

    # Note: requests to the Moxfield API must be rate limited to one request per second.
    @classmethod
    @ratelimit.sleep_and_retry  # type: ignore  # `ratelimit` does not implement decorator typing correctly
    @ratelimit.limits(calls=1, period=1)  # type: ignore  # `ratelimit` does not implement decorator typing correctly
    def retrieve_card_list(cls, url: str) -> str:
        path = urlparse(url).path
        deck_id = path.split("/")[-1]
        response = cls.request(
            path=f"v2/decks/all/{deck_id}", netloc="api.moxfield.com", headers={"User-Agent": settings.MOXFIELD_SECRET}
        )
        response_json = response.json()
        card_list = ""
        for category in [
            "commanders",
            "companions",
            "mainboard",
            "sideboard",
            "maybeboard",
        ]:
            for name, info in response_json.get(category, {}).items():
                card_list += f"{info['quantity']} {name}\n"
        for token in response_json.get("tokens", []):
            if token["layout"] == "token":
                card_list += f"t:{token['name']}\n"
        return card_list


class MTGGoldfish(ImportSite):
    @staticmethod
    def get_host_names() -> list[str]:
        return ["www.mtggoldfish.com", "mtggoldfish.com"]  # mtggoldfish prefers www.

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        path = urlparse(url).path
        deck_id = path.split("#")[0].split("/")[-1]
        if not deck_id:
            raise cls.InvalidURLException(url)
        response = cls.request(path=f"deck/download/{deck_id}")
        return response.text


class Scryfall(ImportSite):
    @staticmethod
    def get_host_names() -> list[str]:
        return ["scryfall.com", "www.scryfall.com"]

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        path = urlparse(url).path
        deck_id = path.rsplit("#", 1)[0].split("/")[-1]
        if not deck_id:
            raise cls.InvalidURLException(url)
        response = cls.request(path=f"decks/{deck_id}/export/text", netloc="api.scryfall.com")
        card_list = response.text
        for line_to_remove in ["// Sideboard"]:
            card_list = card_list.replace(line_to_remove, "")
        return card_list


class TappedOut(ImportSite):
    @staticmethod
    def get_host_names() -> list[str]:
        return ["tappedout.net", "www.tappedout.net"]

    @classmethod
    def retrieve_card_list(cls, url: str) -> str:
        path = urlparse(url).path
        response = cls.request(path=f"{path}?fmt=txt")
        card_list = response.content.decode("utf-8")
        for line_to_remove in ["Sideboard:\r\n"]:
            card_list = card_list.replace(line_to_remove, "")
        return card_list


# endregion


class MTG(GameIntegration):
    """
    Our Magic: The Gathering integration reads DFC pairs from Scryfall and enables reading decklists from some
    popular deckbuilding sites.
    """

    DFC_SCRYFALL_QUERY = "is:dfc -layout:art_series -(layout:double_faced_token -keyword:transform) -is:reversible"
    MELD_SCRYFALL_QUERY = "is:meld"
    DFC_SCRYFALL_URL = f"https://api.scryfall.com/cards/search?q={DFC_SCRYFALL_QUERY}"
    MELD_SCRYFALL_URL = f"https://api.scryfall.com/cards/search?q={MELD_SCRYFALL_QUERY}"

    @classmethod
    def query_scryfall_paginated(cls, url: str) -> list[dict[str, Any]]:
        response = get_json_endpoint_rate_limited(url)
        data = response["data"]
        while response["has_more"]:
            response = get_json_endpoint_rate_limited(response["next_page"])
            data += response["data"]
        return data

    @classmethod
    def get_double_faced_card_pairs(cls) -> list[DFCPair]:
        # query data and construct objects for regular double-faced cards
        dfc_pairs: list[DFCPair] = []
        dfc_data = cls.query_scryfall_paginated(cls.DFC_SCRYFALL_URL)
        print(f"Identified {len(dfc_data)} double-faced cards")
        for item in dfc_data:
            if item["digital"] is True:
                continue
            front_name = item["card_faces"][0]["name"]
            back_name = item["card_faces"][1]["name"]
            dfc_pairs.append(DFCPair(front=front_name, back=back_name))
        return dfc_pairs

    @classmethod
    def get_meld_pairs(cls) -> list[DFCPair]:
        # query data and construct objects for meld pairs

        dfc_pairs: list[DFCPair] = []
        meld_data = cls.query_scryfall_paginated(cls.MELD_SCRYFALL_URL)
        print(f"Identified {len(meld_data)} meld pieces")
        for item in meld_data:
            if "all_parts" not in item:
                card_name = item["name"]
                print(f"Skipping {card_name} (missing key 'all_parts')")
                continue

            card_part_singleton_list = list(filter(lambda part: part["name"] == item["name"], item["all_parts"]))
            meld_result_singleton_list = list(
                filter(lambda part: part["component"] == "meld_result", item["all_parts"])
            )

            if len(card_part_singleton_list) != 1 or len(meld_result_singleton_list) != 1:
                continue

            card_part = card_part_singleton_list.pop()
            meld_result = meld_result_singleton_list.pop()["name"]
            if card_part["component"] == "meld_part":
                is_top = "\n(Melds with " not in item["oracle_text"]
                card_bit = "Top" if is_top else "Bottom"
                dfc_pairs.append(DFCPair(front=item["name"], back=f"{meld_result} {card_bit}"))

        return dfc_pairs

    # region implementation of abstract methods

    @classmethod
    def get_dfc_pairs(cls) -> list[DFCPair]:
        return cls.get_double_faced_card_pairs() + cls.get_meld_pairs()

    @classmethod
    def get_import_sites(cls) -> list[Type[ImportSite]]:
        return [
            Aetherhub,
            Archidekt,
            CubeCobra,
            Deckstats,
            MagicVille,
            ManaStack,
            *([Moxfield] if settings.MOXFIELD_SECRET else []),
            MTGGoldfish,
            Scryfall,
            TappedOut,
        ]

    # endregion


__all__ = ["MTG"]
