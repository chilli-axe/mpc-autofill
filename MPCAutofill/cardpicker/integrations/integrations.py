from typing import Optional, Type

from django.conf import settings

from cardpicker.integrations.game.base import GameIntegration
from cardpicker.integrations.game.mtg import MTG


def get_configured_game_integration() -> Optional[Type[GameIntegration]]:
    integrations: dict[str, Type[GameIntegration]] = {integration.__name__: integration for integration in [MTG]}
    return integrations.get(settings.GAME) if settings.GAME else None
