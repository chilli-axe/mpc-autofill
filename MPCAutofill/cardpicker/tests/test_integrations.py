import time
from collections import Counter
from concurrent.futures import ThreadPoolExecutor
from enum import Enum
from typing import Optional

import pytest
import requests_mock

from django.conf import settings as conf_settings

from cardpicker.integrations.game.mtg import MTG, Moxfield
from cardpicker.integrations.integrations import get_configured_game_integration


class TestGetIntegration:
    @pytest.mark.parametrize("environment_variable, integration_class", [("MTG", MTG)])
    def test_get_integration(self, db, environment_variable, integration_class, settings):
        settings.GAME = environment_variable
        assert get_configured_game_integration() == integration_class


class TestMTGIntegration:
    # region constants

    class Decks(Enum):
        # all of these decks have 4x brainstorm, 3x past in flames, and 1x delver of secrets // insectile aberration
        AETHERHUB = "https://aetherhub.com/Deck/test-796905"
        AETHERHUB_WITH_WWW = "https://www.aetherhub.com/Deck/test-796905"
        ARCHIDEKT = "https://archidekt.com/decks/3380653"
        ARCHIDEKT_WITH_WWW = "https://www.archidekt.com/decks/3380653"
        ARCHIDEKT_WITH_HASH = "https://archidekt.com/decks/3380653#test"
        ARCHIDEKT_WITH_SLASH = "https://archidekt.com/decks/3380653/test"
        CUBECOBRA = "https://cubecobra.com/cube/overview/2fj4"
        CUBECOBRA_WITH_WWW = "https://www.cubecobra.com/cube/overview/2fj4"
        DECKSTATS = "https://deckstats.net/decks/216625/2754468-test"
        DECKSTATS_WITH_WWW = "https://www.deckstats.net/decks/216625/2754468-test"
        MAGICVILLE = "https://magic-ville.com/fr/decks/showdeck?ref=948045"
        MAGICVILLE_WITH_WWW = "https://www.magic-ville.com/fr/decks/showdeck?ref=948045"
        MANASTACK = "https://manastack.com/deck/test-426"
        MOXFIELD = "https://www.moxfield.com/decks/D42-or9pCk-uMi4XzRDziQ"
        MOXFIELD_WITHOUT_WWW = "https://moxfield.com/decks/D42-or9pCk-uMi4XzRDziQ"
        MTGGOLDFISH = "https://www.mtggoldfish.com/deck/5149750"
        MTGGOLDFISH_WITHOUT_WWW = "https://mtggoldfish.com/deck/5149750"
        SCRYFALL = "https://scryfall.com/@mpcautofill/decks/71bb2d40-c922-4a01-a63e-7ba2dde29a5c"
        SCRYFALL_WITH_WWW = "https://www.scryfall.com/@mpcautofill/decks/71bb2d40-c922-4a01-a63e-7ba2dde29a5c"
        TAPPEDOUT = "https://tappedout.net/mtg-decks/09-10-22-DoY-test"
        TAPPEDOUT_WITH_WWW = "https://www.tappedout.net/mtg-decks/09-10-22-DoY-test"

    # endregion

    # region fixtures

    @pytest.fixture()
    def moxfield_secret_setter(self):
        def _callable(moxfield_secret: Optional[str]):
            conf_settings.MOXFIELD_SECRET = moxfield_secret

        old_secret = conf_settings.MOXFIELD_SECRET
        yield _callable
        conf_settings.MOXFIELD_SECRET = old_secret

    # endregion

    # region tests

    def test_get_double_faced_card_pairs(self):
        assert len(MTG.get_double_faced_card_pairs()) > 0

    def test_get_meld_pairs(self):
        assert len(MTG.get_meld_pairs()) > 0

    @pytest.mark.parametrize("url", [item.value for item in Decks], ids=[item.name.lower() for item in Decks])
    def test_valid_url(self, client, django_settings, snapshot, url: str):
        decklist = MTG.query_import_site(url)
        assert decklist
        assert Counter(decklist.splitlines()) == snapshot

    @pytest.mark.parametrize(
        "moxfield_secret, is_moxfield_enabled",
        [
            (None, False),
            ("", False),
            ("lorem ipsum", True),
        ],
    )
    def test_moxfield_enabled(self, moxfield_secret_setter, moxfield_secret, is_moxfield_enabled):
        moxfield_secret_setter(moxfield_secret)
        import_sites = MTG.get_import_sites()
        if is_moxfield_enabled:
            assert Moxfield in import_sites
        else:
            assert Moxfield not in import_sites

    def test_moxfield_rate_limit(self, monkeypatch):
        with requests_mock.Mocker() as mock:
            mock.get("https://api.moxfield.com/v2/decks/all/D42-or9pCk-uMi4XzRDziQ", json={})

            with ThreadPoolExecutor(max_workers=3) as pool:
                t0 = time.time()
                pool.map(lambda _: [Moxfield.retrieve_card_list(self.Decks.MOXFIELD.value) for _ in range(2)], range(3))
            t1 = time.time()
            t = t1 - t0
            assert t > 5  # one second between calls

    # endregion
