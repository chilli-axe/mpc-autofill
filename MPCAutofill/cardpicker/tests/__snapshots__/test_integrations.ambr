# name: TestMTGIntegration.test_valid_url[aetherhub]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[aetherhub_with_www]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[archidekt_with_hash]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[archidekt_with_slash]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[archidekt_with_www]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[archidekt]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[cubecobra]
  Counter({
    'Brainstorm': 4,
    'Delver of Secrets': 1,
    'Past in Flames': 3,
  })
# ---
# name: TestMTGIntegration.test_valid_url[cubecobra_with_www]
  Counter({
    'Brainstorm': 4,
    'Delver of Secrets': 1,
    'Past in Flames': 3,
  })
# ---
# name: TestMTGIntegration.test_valid_url[deckstats]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[deckstats_with_www]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[magicville]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[magicville_with_www]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[manastack]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[scryfall]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[scryfall_with_www]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[tappedout]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[tappedout_with_www]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[moxfield]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
    't:Innistrad Checklist': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[moxfield_without_www]
  Counter({
    '1 Delver of Secrets // Insectile Aberration': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
    't:Innistrad Checklist': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[mtggoldfish]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
# name: TestMTGIntegration.test_valid_url[mtggoldfish_without_www]
  Counter({
    '1 Delver of Secrets': 1,
    '3 Past in Flames': 1,
    '4 Brainstorm': 1,
  })
# ---
